File: python3/day09/day09.py
numbers = [int(l) for l in open("input.txt").readlines()]


def part1():
    for i in range(25, len(numbers)):
        sliding_window = numbers[(i - 25) : i]

        found_two_numbers = False
        for j in range(25):
            for k in range(25):
                if (j != k) and (sliding_window[j] + sliding_window[k] == numbers[i]):
                    found_two_numbers = True
                    break

        if not found_two_numbers:
            return numbers[i], i


def part2(outlier, pos):
    for i in range(pos - 2):
        for j in range(i + 2, pos):
            slice = numbers[i:j]
            if sum(slice) == outlier:
                return min(slice) + max(slice)


outlier, pos = part1()
print(outlier)
print(part2(outlier, pos))

File: scala/src/main/scala/year2020/Day09.scala
package com.ochronus
package year2020

import scala.collection.Searching._
import utils.File

class Day09 {
  private val numbers = File.readFileLines("day09.input.txt").map(_.toLong)
  val windowSize = 25

  def part1(): Long = {
      numbers.drop(windowSize).zip(numbers.sliding(windowSize)).find(elem => elem._2.combinations(2).forall(_.sum != elem._1)).get._1
  }

  def part2(outlier: Long): Long = {
    for (i <- 0 to numbers.length - 1) {
      for (j <- i + 2 to numbers.length - 1) {
        val slice = numbers.drop(i).take(j - i)
        if (slice.sum == outlier) return (slice.min + slice.max)
      }
    }
    0
  }
}
